cheroot/test/test_core.py

```diff
@@ -351,3 +351,56 @@ def test_garbage_in(test_client):
         # "Connection reset by peer" is also acceptable.
         if ex.errno != errno.ECONNRESET:
             raise
+
+
+class CloseController(object):
+    """Controller for testing the close callback."""
+
+    def __init__(self):
+        """Sets the default values."""
+        self.sent_headers_before_closing = None
+        self.req = None
+
+    def __call__(self, environ, start_response):
+        """Gets the req to know header sent status."""
+        self.req = start_response.__self__.req
+        resp = CloseResponse(self.close)
+        start_response(resp.status, resp.headers.items())
+        return resp
+
+    def close(self):
+        """Hook for close, tests if headers already sent."""
+        self.sent_headers_before_closing = \
+            self.req is not None and self.req.sent_headers
+
+
+class CloseResponse(object):
+    """Dummy empty response to trigger the no body status."""
+    def __init__(self, close):
+        """Use some defaults to ensure we have a header."""
+        self.status = '200 OK'
+        self.headers = {'Content-Type': 'text/html'}
+        self.close = close
+
+    def __getitem__(self, index):
+        """Ensures we don't have a body."""
+        raise IndexError()
+
+    def output(self):
+        """Returns self to hook the close method."""
+        return self
+
+
+@pytest.fixture
+def testing_server_close(wsgi_server_client):
+    wsgi_server = wsgi_server_client.server_instance
+    wsgi_server.wsgi_app = CloseController()
+    wsgi_server.max_request_body_size = 30000000
+    wsgi_server.server_client = wsgi_server_client
+    return wsgi_server
+
+
+def test_send_header_before_closing(testing_server_close):
+    """Tests we are actually sending the headers before closing the response."""
+    testing_server_close.server_client.get('/')
```

Comment:
How about checking http response headers in client?
```python
resp_status_line, resp_headers, resp_body = testing_server_close.server_client.get('/')
```

Reply:
Sorry, maybe my description is wrong here or i didn't understand well.

The headers are always sent, the problem is that the `close` [1] method is being call before the headers are being sent.

This is a problem in cherrypy, because the `on_end_request` is being called before the headers are actually sent, anything sent to the client from this hook will reach before the headers, thus the headers will be malformed. I don't think I can test this from the client side unless i can get a raw response, but i would still need to send something on that `close` [1] method.

It's done in a similar fashion in pep-0333 [2] 

```python
    try:
        for data in result:
            if data:    # don't send headers until body appears
                write(data)
        if not headers_sent:
            write('')   # send headers now if body was empty
    finally:
        if hasattr(result, 'close'):
            result.close()

```

[1] https://github.com/cherrypy/cheroot/pull/75/files#diff-08d927a6f63bcf494ed88cca2a6873d0R144
[2] https://www.python.org/dev/peps/pep-0333/